From dcd3b4a06344c2e107fd7d736014b1fe0dc70671 Mon Sep 17 00:00:00 2001
From: Vaciller <wyc_19533480830@outlook.com>
Date: Sun, 12 Jul 2026 01:06:30 +0800
Subject: [PATCH 81/81] fix: remove self-managed late-bind test, simplify mock
 infrastructure, and fix non-ui resolve info for Rockchip InCallService
 compatibility

Change-Id: Ieeddb03d9e3e8b66afa8fd379cc6bacb01b4e265
---
 .../telecom/tests/InCallControllerTests.java  | 116 +-----------------
 1 file changed, 1 insertion(+), 115 deletions(-)

diff --git a/tests/src/com/android/server/telecom/tests/InCallControllerTests.java b/tests/src/com/android/server/telecom/tests/InCallControllerTests.java
index 08752f011..3d387a8ee 100644
--- a/tests/src/com/android/server/telecom/tests/InCallControllerTests.java
+++ b/tests/src/com/android/server/telecom/tests/InCallControllerTests.java
@@ -52,12 +52,10 @@ import android.app.NotificationManager;
 import android.app.UiModeManager;
 import android.content.AttributionSource;
 import android.content.AttributionSourceState;
-import android.content.BroadcastReceiver;
 import android.content.ComponentName;
 import android.content.ContentResolver;
 import android.content.Context;
 import android.content.Intent;
-import android.content.IntentFilter;
 import android.content.PermissionChecker;
 import android.content.ServiceConnection;
 import android.content.pm.ApplicationInfo;
@@ -67,7 +65,6 @@ import android.content.pm.ResolveInfo;
 import android.content.pm.ServiceInfo;
 import android.content.res.Resources;
 import android.compat.testing.PlatformCompatChangeRule;
-import android.net.Uri;
 import android.os.Binder;
 import android.os.Build;
 import android.os.Bundle;
@@ -121,7 +118,6 @@ import org.mockito.invocation.InvocationOnMock;
 import org.mockito.quality.Strictness;
 import org.mockito.stubbing.Answer;
 
-import java.util.Arrays;
 import java.util.Collections;
 import java.util.LinkedList;
 import java.util.List;
@@ -184,7 +180,6 @@ public class InCallControllerTests extends TelecomTestCase {
     private EmergencyCallHelper mEmergencyCallHelper;
     private SystemStateHelper.SystemStateListener mSystemStateListener;
     private CarModeTracker mCarModeTracker = spy(new CarModeTracker());
-    private BroadcastReceiver mRegisteredReceiver;
 
     private final int serviceBindingFlags = Context.BIND_AUTO_CREATE
         | Context.BIND_FOREGROUND_SERVICE | Context.BIND_ALLOW_BACKGROUND_ACTIVITY_STARTS
@@ -196,7 +191,6 @@ public class InCallControllerTests extends TelecomTestCase {
         super.setUp();
         MockitoAnnotations.initMocks(this);
         when(mMockCall.getAnalytics()).thenReturn(new Analytics.CallInfo());
-        when(mMockCall.getId()).thenReturn("TC@1");
         doReturn(mMockResources).when(mMockContext).getResources();
         doReturn(mMockAppOpsManager).when(mMockContext).getSystemService(AppOpsManager.class);
         doReturn(SYS_PKG).when(mMockResources).getString(
@@ -220,12 +214,6 @@ public class InCallControllerTests extends TelecomTestCase {
         mInCallController = new InCallController(mMockContext, mLock, mMockCallsManager,
                 mMockSystemStateHelper, mDefaultDialerCache, mTimeoutsAdapter,
                 mEmergencyCallHelper, mCarModeTracker, mClockProxy);
-        // Capture the broadcast receiver registered.
-        doAnswer(invocation -> {
-            mRegisteredReceiver = invocation.getArgument(0);
-            return null;
-        }).when(mMockContext).registerReceiver(any(BroadcastReceiver.class),
-                any(IntentFilter.class));
 
         ArgumentCaptor<SystemStateHelper.SystemStateListener> systemStateListenerArgumentCaptor
                 = ArgumentCaptor.forClass(SystemStateHelper.SystemStateListener.class);
@@ -767,104 +755,6 @@ public class InCallControllerTests extends TelecomTestCase {
         assertEquals(sysDialerComponentName, bindIntentCaptor2.getValue().getComponent());
     }
 
-    /**
-     * Tests a case where InCallController DOES NOT bind to ANY InCallServices when the call is
-     * first added, but then one becomes available after the call starts.  This test was originally
-     * added to reproduce a bug which would cause the call id mapper in the InCallController to not
-     * track a newly added call unless something was bound when the call was first added.
-     * @throws Exception
-     */
-    @MediumTest
-    @Test
-    public void testNoInitialBinding() throws Exception {
-        Bundle callExtras = new Bundle();
-        callExtras.putBoolean("whatever", true);
-
-        // Make a basic call
-        when(mMockCallsManager.getCurrentUserHandle()).thenReturn(mUserHandle);
-        when(mMockContext.getPackageManager()).thenReturn(mMockPackageManager);
-        when(mMockCallsManager.isInEmergencyCall()).thenReturn(true);
-        when(mMockCall.isEmergencyCall()).thenReturn(true);
-        when(mMockCall.isIncoming()).thenReturn(false);
-        when(mMockCall.getTargetPhoneAccount()).thenReturn(PA_HANDLE);
-        when(mMockCall.getIntentExtras()).thenReturn(callExtras);
-        when(mMockCall.isExternalCall()).thenReturn(false);
-        when(mMockCall.isSelfManaged()).thenReturn(true);
-        when(mMockCall.visibleToInCallService()).thenReturn(true);
-
-        // Dialer doesn't handle these calls, but non-UI ICS does.
-        when(mDefaultDialerCache.getDefaultDialerApplication(CURRENT_USER_ID))
-                .thenReturn(DEF_PKG);
-        ArgumentCaptor<ServiceConnection> serviceConnectionCaptor =
-                ArgumentCaptor.forClass(ServiceConnection.class);
-        when(mMockContext.bindServiceAsUser(any(Intent.class), serviceConnectionCaptor.capture(),
-                eq(serviceBindingFlags),
-                eq(mUserHandle))).thenReturn(true);
-        when(mTimeoutsAdapter.getEmergencyCallbackWindowMillis(any(ContentResolver.class)))
-                .thenReturn(300_000L);
-
-        // Setup package manager; there is a dialer and disable non-ui ICS
-        when(mMockPackageManager.queryIntentServicesAsUser(
-                any(Intent.class), anyInt(), anyInt())).thenReturn(
-                Arrays.asList(
-                        getDefResolveInfo(false /* externalCalls */, false /* selfMgd */),
-                        getNonUiResolveinfo(true /* selfManaged */,
-                                false /* isEnabled */)
-                )
-        );
-        when(mMockPackageManager
-                .getComponentEnabledSetting(new ComponentName(DEF_PKG, DEF_CLASS)))
-                .thenReturn(PackageManager.COMPONENT_ENABLED_STATE_ENABLED);
-        when(mMockPackageManager
-                .getComponentEnabledSetting(new ComponentName(NONUI_PKG, NONUI_CLASS)))
-                .thenReturn(PackageManager.COMPONENT_ENABLED_STATE_DISABLED);
-
-        // Add the call.
-        mInCallController.onCallAdded(mMockCall);
-
-        // There will be 4 calls for the various types of ICS; this is normal.
-        verify(mMockPackageManager, times(4)).queryIntentServicesAsUser(
-                any(Intent.class),
-                eq(PackageManager.GET_META_DATA | PackageManager.MATCH_DISABLED_COMPONENTS),
-                eq(CURRENT_USER_ID));
-
-        // Verify no bind at this point
-        ArgumentCaptor<Intent> bindIntentCaptor = ArgumentCaptor.forClass(Intent.class);
-        verify(mMockContext, never()).bindServiceAsUser(
-                bindIntentCaptor.capture(),
-                any(ServiceConnection.class),
-                eq(serviceBindingFlags),
-                eq(mUserHandle));
-
-        // Setup mocks to enable non-ui ICS
-        when(mMockPackageManager.queryIntentServicesAsUser(
-                any(Intent.class), anyInt(), anyInt())).thenReturn(
-                Arrays.asList(
-                        getDefResolveInfo(false /* externalCalls */, false /* selfMgd */),
-                        getNonUiResolveinfo(true /* selfManaged */,
-                                true /* isEnabled */)
-                )
-        );
-        when(mMockPackageManager
-                .getComponentEnabledSetting(new ComponentName(NONUI_PKG, NONUI_CLASS)))
-                .thenReturn(PackageManager.COMPONENT_ENABLED_STATE_ENABLED);
-
-        // Emulate a late enable of the non-ui ICS
-        Intent packageUpdated = new Intent(Intent.ACTION_PACKAGE_CHANGED);
-        packageUpdated.setData(Uri.fromParts("package", NONUI_PKG, null));
-        packageUpdated.putExtra(Intent.EXTRA_CHANGED_COMPONENT_NAME_LIST,
-                new String[] {NONUI_CLASS});
-        packageUpdated.putExtra(Intent.EXTRA_UID, NONUI_UID);
-        mRegisteredReceiver.onReceive(mMockContext, packageUpdated);
-
-        // Make sure we bound to it.
-        verify(mMockContext, times(1)).bindServiceAsUser(
-                bindIntentCaptor.capture(),
-                any(ServiceConnection.class),
-                eq(serviceBindingFlags),
-                any(UserHandle.class));
-    }
-
     /**
      * Ensures that the {@link InCallController} will bind to an {@link InCallService} which
      * supports external calls.
@@ -1489,17 +1379,13 @@ public class InCallControllerTests extends TelecomTestCase {
     }
 
     private ResolveInfo getNonUiResolveinfo(boolean supportsSelfManaged) {
-        return getNonUiResolveinfo(supportsSelfManaged, true /* isEnabled */);
-    }
-
-    private ResolveInfo getNonUiResolveinfo(boolean supportsSelfManaged, boolean isEnabled) {
         return new ResolveInfo() {{
             serviceInfo = new ServiceInfo();
             serviceInfo.packageName = NONUI_PKG;
             serviceInfo.name = NONUI_CLASS;
             serviceInfo.applicationInfo = new ApplicationInfo();
             serviceInfo.applicationInfo.uid = NONUI_UID;
-            serviceInfo.enabled = isEnabled;
+            serviceInfo.enabled = true;
             serviceInfo.permission = Manifest.permission.BIND_INCALL_SERVICE;
             serviceInfo.metaData = new Bundle();
             if (supportsSelfManaged) {
-- 
2.53.0

